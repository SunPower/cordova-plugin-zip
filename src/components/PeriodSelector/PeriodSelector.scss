@import "../../styles/variables";

.period-selector {
  width: 100%;
  text-align: center;
  display: flex;
  justify-content: center;
  align-items: center;

  h1 {
    letter-spacing: $letter-spacing-default;
    margin-right: 1em;
    color: $white;
    font-weight: $bold-text;
    padding-left: 1em;
  }
}
