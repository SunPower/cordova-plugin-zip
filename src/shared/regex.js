export const PASSWORD_REGEXP = /^(?=.*[a-z])(?=.*\d)[A-Za-z\d]{8,}$/
export const EMAIL_REGEXP = /^\S+@\S+\.\S+$/
export const PHONE_REGEXP = /^(1\s?)?((([0-9]{3}))|[0-9]{3})[\s-]?[0-9]{3}[\s-]?[0-9]{4}$/
