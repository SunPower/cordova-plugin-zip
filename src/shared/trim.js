export const trimObject = obj => {
  return Object.entries(obj)
    .map(([k, v]) => ({
      [k]: typeof v === 'string' || v instanceof String ? v.trim() : v
    }))
    .reduce((acc, kv) => ({ ...acc, ...kv }), {})
}
