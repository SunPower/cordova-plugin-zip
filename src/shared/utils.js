export const either = (condition, whenTrue, whenFalse) =>
  condition ? whenTrue : whenFalse
