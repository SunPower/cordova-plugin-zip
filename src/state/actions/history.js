import { createAction } from 'redux-act'

export const CHANGE_INTERVAL_INIT = createAction('CHANGE_INTERVAL_INIT')
export const CHANGE_INTERVAL = createAction('CHANGE_INTERVAL')
export const CHANGE_INTERVAL_DELTA = createAction('CHANGE_INTERVAL_DELTA')
export const CHANGE_INTERVAL_SUCCESS = createAction('CHANGE_INTERVAL_SUCCESS')
export const CHANGE_INTERVAL_ERROR = createAction('CHANGE_INTERVAL_ERROR')

export const INTERVAL_DELTA = {
  DAY: 'DAY',
  WEEK: 'WEEK',
  MONTH: 'MONTH',
  YEAR: 'YEAR',
  LIFETIME: 'LIFETIME',
  CUSTOM: 'CUSTOM'
}

export const changeDatePeriod = datePeriod => {
  return dispatch => {
    try {
      dispatch(CHANGE_INTERVAL_INIT())
      dispatch(CHANGE_INTERVAL_DELTA(datePeriod))
      dispatch(CHANGE_INTERVAL_SUCCESS())
    } catch (err) {
      dispatch(CHANGE_INTERVAL_ERROR(err))
    }
  }
}
