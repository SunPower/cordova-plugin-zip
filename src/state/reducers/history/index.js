import { createReducer } from 'redux-act'
import moment from 'moment'
import {
  CHANGE_INTERVAL_INIT,
  CHANGE_INTERVAL,
  CHANGE_INTERVAL_DELTA,
  CHANGE_INTERVAL_SUCCESS
} from '../../actions/history'

const currentDate = moment()
const initialState = {
  intervalDelta: 'DAY',
  interval: [
    moment(currentDate).startOf('day'),
    moment(currentDate).endOf('day')
  ],
  changingInterval: false
}

export const historyReducer = createReducer(
  {
    [CHANGE_INTERVAL_INIT]: state => ({
      ...state,
      changingInterval: true
    }),
    [CHANGE_INTERVAL]: (state, { interval }) => ({
      ...state,
      interval
    }),
    [CHANGE_INTERVAL_DELTA]: (state, intervalDelta) => ({
      ...state,
      intervalDelta: intervalDelta
    }),
    [CHANGE_INTERVAL_SUCCESS]: state => ({
      ...state,
      changingInterval: false
    })
  },
  initialState
)
