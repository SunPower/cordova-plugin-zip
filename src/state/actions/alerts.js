import { createAction } from 'redux-act'
import { httpGet } from '../../shared/fetch'

export const ALERTS_FETCH_INIT = createAction('ALERTS_FETCH_INIT')
export const ALERTS_FETCH_SUCCESS = createAction('ALERTS_FETCH_SUCCESS')
export const ALERTS_FETCH_ERROR = createAction('ALERTS_FETCH_ERROR')

let alertsPollTimer
const ALERTS_POLL_INTERVAL = 1000 * 60

export const pollAlerts = () => {
  return (dispatch, getState) => {
    const callback = async () => {
      const state = getState()
      const isLoggedIn = !!state.user.auth.tokenID

      if (alertsPollTimer && !isLoggedIn) {
        clearInterval(alertsPollTimer)
        return
      }

      const addressId = state.user.auth.addressId

      try {
        dispatch(ALERTS_FETCH_INIT())
        const { status, data } = await httpGet(
          `/address/${addressId}/alerts`,
          state
        )
        return status === 200
          ? dispatch(ALERTS_FETCH_SUCCESS(data))
          : dispatch(ALERTS_FETCH_ERROR(data))
      } catch (err) {
        dispatch(ALERTS_FETCH_ERROR(err))
      }
    }

    if (!alertsPollTimer) {
      callback()
      alertsPollTimer = setInterval(callback, ALERTS_POLL_INTERVAL)
    }
  }
}
export const ALERTS_STOP_POLLING_INIT = createAction('ALERTS_STOP_POLLING_INIT')
export const ALERTS_STOP_POLLING_SUCCESS = createAction(
  'ALERTS_STOP_POLLING_SUCCESS'
)

export const stopPollingAlerts = () => {
  return dispatch => {
    dispatch(ALERTS_STOP_POLLING_INIT())
    if (alertsPollTimer) {
      clearInterval(alertsPollTimer)
      dispatch(ALERTS_STOP_POLLING_SUCCESS())
      return
    }
  }
}

export const ALERTS_DISMISS = createAction('ALERTS_DISMISS')
export const ALERTS_SEEN = createAction('ALERTS_SEEN')
