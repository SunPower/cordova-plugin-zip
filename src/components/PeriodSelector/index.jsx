import React from 'react'
import { useI18n } from '../../shared/i18n'
import { useDispatch } from 'react-redux'
import { changeDatePeriod, INTERVAL_DELTA } from '../../state/actions/history'
import PeriodSelectorMenu from '../PeriodSelectorMenu'

import './PeriodSelector.scss'

function changeDatePeriodOption(dispatch) {
  return option => dispatch(changeDatePeriod(option.value))
}

function PeriodSelector({ intervalDelta }) {
  const t = useI18n()
  const dispatch = useDispatch()

  const options = [
    { id: 1, text: t('HISTORY_DAY_VIEW'), value: INTERVAL_DELTA.DAY },
    { id: 2, text: t('HISTORY_WEEK_VIEW'), value: INTERVAL_DELTA.WEEK },
    { id: 3, text: t('HISTORY_MONTH_VIEW'), value: INTERVAL_DELTA.MONTH },
    {
      id: 4,
      text: t('HISTORY_YEAR_VIEW'),
      value: INTERVAL_DELTA.YEAR,
      disabled: true
    },
    {
      id: 5,
      text: t('HISTORY_LIFETIME_VIEW'),
      value: INTERVAL_DELTA.LIFETIME,
      disabled: true
    },
    {
      id: 6,
      text: t('HISTORY_CUSTOM_DATES_VIEW'),
      value: INTERVAL_DELTA.CUSTOM,
      disabled: true
    }
  ]

  const selectedOpt = options.find(({ value }) => value === intervalDelta) || {}
  selectedOpt.selected = true

  return (
    <PeriodSelectorMenu
      header={t('HISTORY_TIMEFRAME_HEADER')}
      options={options}
      className="is-seven-eighths"
      onClick={changeDatePeriodOption(dispatch)}
      modalId="modal-change-timeframe"
    />
  )
}

export default PeriodSelector
