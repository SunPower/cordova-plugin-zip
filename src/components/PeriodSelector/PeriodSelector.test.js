import React from 'react'
import { shallow } from 'enzyme'
import PeriodSelector from '.'
import * as i18n from '../../shared/i18n'
import * as reactRedux from 'react-redux'
import * as historyActions from '../../state/actions/history'
import { changeDatePeriod } from '../../state/actions/history'

describe('History View Period Selector component', () => {
  let dispatchMock

  beforeEach(() => {
    dispatchMock = jest.fn()
    jest
      .spyOn(i18n, 'useI18n')
      .mockImplementation(path => (key, ...params) =>
        `${key.toUpperCase()} ${params.join('_')}`.trim()
      )
    jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => dispatchMock)
    jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => dispatchMock)
    jest.spyOn(historyActions, 'changeDatePeriod').mockImplementation(jest.fn)
  })

  it('renders without crashing', () => {
    const component = shallow(<PeriodSelector />)
    expect(component).toMatchSnapshot()
  })

  it('changes timeframe successfully', () => {
    const modalRoot = global.document.createElement('div')
    modalRoot.setAttribute('id', 'modal-root')

    const body = global.document.querySelector('body')
    body.appendChild(modalRoot)

    const { component } = mountWithProvider(<PeriodSelector />)({})
    component
      .find('.chevronDown')
      .at(0)
      .simulate('click')

    component
      .find('.date-option')
      .at(0)
      .simulate('click')
    expect(changeDatePeriod).toBeCalled()
  })
})
