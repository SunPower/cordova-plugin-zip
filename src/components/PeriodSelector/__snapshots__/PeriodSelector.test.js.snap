// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History View Period Selector component renders without crashing 1`] = `
<PeriodSelectorMenu
  className="is-seven-eighths"
  header="HISTORY_TIMEFRAME_HEADER"
  modalId="modal-change-timeframe"
  onClick={[Function]}
  options={
    Array [
      Object {
        "id": 1,
        "text": "HISTORY_DAY_VIEW",
        "value": "DAY",
      },
      Object {
        "id": 2,
        "text": "HISTORY_WEEK_VIEW",
        "value": "WEEK",
      },
      Object {
        "id": 3,
        "text": "HISTORY_MONTH_VIEW",
        "value": "MONTH",
      },
      Object {
        "disabled": true,
        "id": 4,
        "text": "HISTORY_YEAR_VIEW",
        "value": "YEAR",
      },
      Object {
        "disabled": true,
        "id": 5,
        "text": "HISTORY_LIFETIME_VIEW",
        "value": "LIFETIME",
      },
      Object {
        "disabled": true,
        "id": 6,
        "text": "HISTORY_CUSTOM_DATES_VIEW",
        "value": "CUSTOM",
      },
    ]
  }
/>
`;
